### ISSUE TITLE
[BC] Change spaces to underscores in permission machine names

### ISSUE THREAD

**Issue by ghost** [backdrop/backdrop-issues#3787]
It seems that permissions are the only things (that I know of) in Backdrop (and even Drupal 7) whose machine names can have spaces in them. Everything else (content types, image styles, etc.) have machine names with just "lowercase alphanumeric characters, underscores (_), and hyphens (-)".

Can/should we change permission machine names to match everything else? Is there a (valid) reason permissions are setup like this? This'd obviously be a major API change, so it'd only go in 2.x, if at all.

**Comment by klonos** [comment]
I agree that this is an odd thing that has bugged me as well. Also agree that it needs to be a 2.x thing.

**Comment by BWPanda** [comment]
Unless... We added something like this to the appropriate function(s):

```
$permission_name = str_replace (' ', '_', $permission_name);
```

**Comment by bugfolder** [comment]
Unifying machine names in that way seems like a worthy goal, and the "have simple rules" gene in me is totally in favor of it, but yeah, it would affect a lot of contrib modules that use permissions from core (or from other contrib modules).

Would this policy be strongly enforced in 2.0 on contrib or custom modules, and if so, how? Currently there's nothing preventing someone from creating a contrib module with a permission `'access MyModule'`. It would seem that there might ought to be a test added to the standard test suite for all contrib modules to ensure that people are following this rule. (Also, of course, documentation added to `hook_permission()`.)

It would also be really nice if we could help users along in this change, by, for example, providing a version of Coder Upgrade that automatically converts non-compliant permissions in 1.x modules.

In fact, might I suggest that for every old-code-breaking change that goes into 2.x, along with the commit(s) that implement the change, there also be commits that (to the extent possible) implement some sort of Coder Upgrade-like funtionality to ease the process for people making their modules 2.x-compliant. (I am a huge fan of CU, and think it needs wider advertisement, even now.)

**Comment by klonos** [comment]
I've been experimenting with something like this on my local (as part of #5030):

```php
function hook_permission() {
  return array(
    'administer my module' => array(
      // @todo: Remove in 2.x.
      'alias' => 'administer_my_module',
    ),
    'administer_my_module' => array(
      'title' => t('Administer my module'),
      'description' => t('Perform administration tasks for my module.'),
    ),
  );
}
```

So basically, we allow permissions definitions to have an `'alias'` key, "redirecting" to a new/renamed permission.

**Comment by bugfolder** [comment]
Is the idea to introduce permission aliases now, to encourage people to use the new form, and then in 2.x, modules would drop the old-style permission and its alias?

**Comment by klonos** [comment]
Yup. If this idea is accepted, we could implement this during the 1.x cycle, instead of having to wait for 2.x.

**Comment by indigoxela** [comment]
> If this idea is accepted, we could implement this during the 1.x cycle

This is a radical change just because someone felt that spaces seem odd. And requires a huge effort, and potentially code bloat to provide backwards compatibility. Sorry, I simply don't get the point. IMO effort and gain are in no reasonable proportion.

To me this looks like a change for 2.x.

**Comment by swampopus** [comment]
Not to muddy the water, but I always hated how in hook_menu, items' elements were defined with spaces in things like:
```php
'page callback' => 'something',
'page arguments' => 'something',
'access callback' => 'something',
```
and so on.  `page_callback` and so forth, with **underscores**, just makes it more consistent and expected.  Also, should the elements begin with `#`, like most of the renderable arrays?

Maybe this should be its own issue.  Or, this issue should be expanded to make everything in a machine-name context use underscores and not spaces.

We could also easily make it to where spaces are replaced by _ automatically, when the menu router cache is rebuilt and when menu items are looked up, and the use of spaces gets a deprecation warning.

**Comment by argiepiano** [comment]
I think this conversation was limited to hook_permission's machine names. I'm not sure the property keys in hook_menu are "machine names" in the sense stated in the title.

Yes, it's odd that several "info" hooks use keys with spaces (like hook_menu and hook_entity_info) and others don't (hook_config_info).  As a side note the `#` in render arrays was a necessity to differentiate the key as "property" of the element rather than as a "child" element. This allows for nesting. I would prefer avoiding `#` in contexts that are not render arrays - there is no need.

**Comment by avpaderno** [comment]
Backdrop allowed spaces in permission machine names for backward compatibility with Drupal 7. Changing them to not accept spaces seems a late change which would not introduce any pro.
